libfoundation: Missing pieces from hash func unification

peter-b opened the pull request:
This PR adds some missing pieces from #5278.

- Clean up some leftover dead code
- Actually implement `MCHashNativeChars()` and `MCHashChars()`
- Add span-based overloads of hash functions

This also adds new `MCHashObject<T>()` and `MCHashObjectStream<T...>()` templates which tidy up pretty much every use of `MCHashBytesStream()` in libfoundation.

peter-b committed:
libfoundation: Remove vestigial "#undef"

The `ELF_STEP` macro was removed in commit b14051d.

peter-b committed:
libfoundation: Add span-based versions of hash functions

peter-b committed:
libfoundation: Use span hashes in hash tests

peter-b committed:
libfoundation: Implement MCHashNativeChars() and MCHashChars()

peter-b committed:
libfoundation: Use object hashes for record hashing.

peter-b committed:
libfoundation: Use object hashes for typeinfo hashing.

peter-b committed:
libfoundation: Add object hashes

There's a fairly common requirement to perform a byte hash on the C++
_object representation_ of a value.  Actual example from libfoundation:

    t_hash = MCHashBytesStream(t_hash, &self->alias.name,
                               sizeof(self->alias.name));

This patch adds two new template functions to libfoundation:
`MCHashObject<T>()` and `MCHashObjectStream<T...>()`.  These functions
take a reference to a value, and return the hash of its object
representation.  `MCHashObjectStream<T...>()` is variadic in order to
allow multiple values of different types to be ergonomically
integrated into a hash.

To ensure correctness, these templates are only enabled for
`TriviallyCopyable` types, i.e. types for which copying the bytes
occupied by the object in the storage is sufficient to produce another
object with the same value.

The example above is then replaceable by:

    t_hash = MCHashObjectStream(t_hash, self->alias.name);

For now, these are only for use within libfoundation due to their
dependence on `<type_traits>`, but they could be moved into the public
API at some point.

runrevmark commented:
@livecode-vulcan review ok 2e1a191

livecode-vulcan commented:
💙 review by @runrevmark ok 2e1a1912fe47a4a1bb02edd3b24430f5ab67bea3

livecode-vulcan commented:
😭 test error 2e1a1912fe47a4a1bb02edd3b24430f5ab67bea3

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1117)
* try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/1119)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1112)
* try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1117)
* try-community-ios: [error](https://vulcan.livecode.com/builders/try-community-ios/builds/1121)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1132)
* try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1113)

peter-b commented:
@livecode-vulcan retry 2e1a191

livecode-vulcan commented:
😭 test error 2e1a1912fe47a4a1bb02edd3b24430f5ab67bea3

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1118)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1120)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1113)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1118)
* try-community-ios: [error](https://vulcan.livecode.com/builders/try-community-ios/builds/1123)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1133)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1114)

peter-b commented:
@livecode-vulcan review not okay 2e1a191

[Bug here](https://github.com/livecode/livecode/pull/5289/commits/2e1a1912fe47a4a1bb02edd3b24430f5ab67bea3#diff-402f0605e2ec70b3bcf5251e3c261ae2R1248)

livecode-vulcan commented:
💔 review by @peter-b not ok 2e1a1912fe47a4a1bb02edd3b24430f5ab67bea3

peter-b committed:
libfoundation: Add span-based versions of object hashes

`MCHashSpan()` and `MCHashSpanStream()` are intended to assist with
hashing a contiguous array of identical objects, e.g. hashing an array
of `int`.

peter-b commented:
@runrevmark I'd accidentally modified the handler and record type info hashes to hash the pointer to their internal field/parameter type info arrays, rather than actually hashing the field/parameter typeinfos themselves.  I've modified this patch to add functions for hashing spans of things.

runrevmark commented:
@livecode-vulcan review ok 8fd70df

livecode-vulcan commented:
💙 review by @runrevmark ok 8fd70df6936044a3cbc82d75df1115e5ef6aa367

livecode-vulcan commented:
😎 test success 8fd70df6936044a3cbc82d75df1115e5ef6aa367

* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/5)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/7)
* try-community-universal-ios-iphoneos10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.2/builds/7)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/7)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/2)
* try-community-universal-ios-iphonesimulator10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.2/builds/7)
* try-community-js-emscripten: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten/builds/9)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/7)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/6)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/1)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/3)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/5)

peter-b commented:
Retrying to check build is correct with updated build service

@livecode-vulcan retry 8fd70df

livecode-vulcan commented:
😎 test success 8fd70df6936044a3cbc82d75df1115e5ef6aa367

* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/5)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/5)
* try-community-js-emscripten: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten/builds/12)
* try-community-universal-ios-iphoneos10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.2/builds/10)
* try-community-universal-ios-iphonesimulator10.2: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.2/builds/10)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/4)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/6)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/8)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/10)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/8)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/10)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/10)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/9)
